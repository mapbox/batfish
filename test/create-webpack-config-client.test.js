'use strict';

const path = require('path');
const createWebpackConfigClient = require('../src/node/create-webpack-config-client');
const createWebpackConfigBase = require('../src/node/create-webpack-config-base');
const validateConfig = require('../src/node/validate-config');
const projectRootSerializer = require('./test-util/project-root-serializer');

jest.mock('../src/node/create-webpack-config-base', () => {
  return jest.fn(() => Promise.resolve({ createdWebpackConfigBase: true }));
});

jest.mock('path-type', () => {
  return {
    dirSync: jest.fn(() => true),
    fileSync: jest.fn(() => true)
  };
});

expect.addSnapshotSerializer(projectRootSerializer);

function createBatfishConfig(options) {
  return validateConfig(
    Object.assign(
      {
        pagesDirectory: path.join(__dirname, './fixtures/get-pages-data')
      },
      options
    )
  );
}

describe('createWebpackConfigClient', () => {
  test('creates a base config', () => {
    const batfishConfig = createBatfishConfig();
    return createWebpackConfigClient(batfishConfig).then((webpackConfig) => {
      expect(createWebpackConfigBase).toHaveBeenCalledTimes(1);
      expect(createWebpackConfigBase).toHaveBeenCalledWith(batfishConfig, {
        target: 'browser'
      });
      expect(webpackConfig.createdWebpackConfigBase).toBe(true);
    });
  });

  test('default Webpack config', () => {
    return createWebpackConfigClient(createBatfishConfig()).then(
      (webpackConfig) => {
        expect(webpackConfig).toMatchSnapshot();
      }
    );
  });

  test('default production Webpack config', () => {
    return createWebpackConfigClient(
      createBatfishConfig({ production: true })
    ).then((webpackConfig) => {
      expect(webpackConfig).toMatchSnapshot();
    });
  });

  test('with all Batfish config options that make a difference', () => {
    const batfishConfig = createBatfishConfig({
      includePromisePolyfill: false,
      vendorModules: ['pigman', 'revenge-of-pigman'],
      outputDirectory: path.join(__dirname, './fake/output/directory'),
      inlineJs: [
        { filename: path.join(__dirname, './sea-creatures.js') },
        { filename: path.join(__dirname, './land-creatures.js') }
      ],
      webpackConfigClientTransform: (x) => {
        x.underwentClientTransform = true;
        return x;
      }
    });
    return createWebpackConfigClient(batfishConfig).then((webpackConfig) => {
      expect(webpackConfig).toMatchSnapshot();
    });
  });
});
