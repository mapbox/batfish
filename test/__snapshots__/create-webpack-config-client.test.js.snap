// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createWebpackConfigClient default Webpack config 1`] = `
Object {
  "createdWebpackConfigBase": true,
  "entry": Object {
    "app": Array [
      "<PROJECT_ROOT>/node_modules/core-js/modules/es6.promise.js",
      "<PROJECT_ROOT>/node_modules/core-js/modules/es6.array.iterator.js",
      "<PROJECT_ROOT>/src/webpack/render-batfish-app.js",
    ],
    "vendor": Array [
      "<PROJECT_ROOT>/node_modules/es6-promise/auto.js",
      "<PROJECT_ROOT>/node_modules/react",
      "<PROJECT_ROOT>/node_modules/react-dom",
      "<PROJECT_ROOT>/node_modules/react-helmet",
      "<PROJECT_ROOT>/node_modules/@mapbox/scroll-restorer/index.js",
      "<PROJECT_ROOT>/node_modules/@mapbox/link-hijacker/index.js",
    ],
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "moduleIds": "hashed",
    "runtimeChunk": Object {
      "name": "manifest",
    },
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendor": Object {
          "chunks": "initial",
          "enforce": true,
          "name": "vendor",
          "test": "vendor",
        },
      },
    },
  },
  "output": Object {
    "chunkFilename": "[name].chunk.js",
    "filename": "[name].js",
  },
  "plugins": Array [
    AssetsWebpackPlugin {
      "options": Object {
        "filename": "assets.json",
        "fullPath": true,
        "manifestFirst": true,
        "path": "<PROJECT_ROOT>/_batfish_site",
        "prettyPrint": false,
        "processOutput": [Function],
        "update": false,
      },
      "writer": [Function],
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.DEV_SERVER": false,
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react": "<PROJECT_ROOT>/node_modules/react",
      "react-dom": "<PROJECT_ROOT>/node_modules/react-dom",
      "react-helmet": "<PROJECT_ROOT>/node_modules/react-helmet",
    },
  },
  "target": "web",
}
`;

exports[`createWebpackConfigClient default production Webpack config 1`] = `
Object {
  "createdWebpackConfigBase": true,
  "entry": Object {
    "app": Array [
      "<PROJECT_ROOT>/node_modules/core-js/modules/es6.promise.js",
      "<PROJECT_ROOT>/node_modules/core-js/modules/es6.array.iterator.js",
      "<PROJECT_ROOT>/src/webpack/render-batfish-app.js",
    ],
    "vendor": Array [
      "<PROJECT_ROOT>/node_modules/es6-promise/auto.js",
      "<PROJECT_ROOT>/node_modules/react",
      "<PROJECT_ROOT>/node_modules/react-dom",
      "<PROJECT_ROOT>/node_modules/react-helmet",
      "<PROJECT_ROOT>/node_modules/@mapbox/scroll-restorer/index.js",
      "<PROJECT_ROOT>/node_modules/@mapbox/link-hijacker/index.js",
    ],
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "moduleIds": "hashed",
    "runtimeChunk": Object {
      "name": "manifest",
    },
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendor": Object {
          "chunks": "initial",
          "enforce": true,
          "name": "vendor",
          "test": "vendor",
        },
      },
    },
  },
  "output": Object {
    "chunkFilename": "[name]-[chunkhash].chunk.js",
    "filename": "[name]-[chunkhash].js",
  },
  "plugins": Array [
    AssetsWebpackPlugin {
      "options": Object {
        "filename": "assets.json",
        "fullPath": true,
        "manifestFirst": true,
        "path": "<PROJECT_ROOT>/_batfish_site",
        "prettyPrint": false,
        "processOutput": [Function],
        "update": false,
      },
      "writer": [Function],
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.DEV_SERVER": false,
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react": "<PROJECT_ROOT>/node_modules/react",
      "react-dom": "<PROJECT_ROOT>/node_modules/react-dom",
      "react-helmet": "<PROJECT_ROOT>/node_modules/react-helmet",
    },
  },
  "target": "web",
}
`;

exports[`createWebpackConfigClient with all Batfish config options that make a difference 1`] = `
Object {
  "createdWebpackConfigBase": true,
  "entry": Object {
    "app": Array [
      "<PROJECT_ROOT>/node_modules/core-js/modules/es6.promise.js",
      "<PROJECT_ROOT>/node_modules/core-js/modules/es6.array.iterator.js",
      "<PROJECT_ROOT>/test/sea-creatures.js",
      "<PROJECT_ROOT>/test/land-creatures.js",
      "<PROJECT_ROOT>/src/webpack/render-batfish-app.js",
    ],
    "vendor": Array [
      "<PROJECT_ROOT>/node_modules/react",
      "<PROJECT_ROOT>/node_modules/react-dom",
      "<PROJECT_ROOT>/node_modules/react-helmet",
      "<PROJECT_ROOT>/node_modules/@mapbox/scroll-restorer/index.js",
      "<PROJECT_ROOT>/node_modules/@mapbox/link-hijacker/index.js",
      "pigman",
      "revenge-of-pigman",
    ],
  },
  "node": Object {
    "child_process": "empty",
    "dgram": "empty",
    "fs": "empty",
    "net": "empty",
    "tls": "empty",
  },
  "optimization": Object {
    "moduleIds": "hashed",
    "runtimeChunk": Object {
      "name": "manifest",
    },
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendor": Object {
          "chunks": "initial",
          "enforce": true,
          "name": "vendor",
          "test": "vendor",
        },
      },
    },
  },
  "output": Object {
    "chunkFilename": "[name].chunk.js",
    "filename": "[name].js",
  },
  "plugins": Array [
    AssetsWebpackPlugin {
      "options": Object {
        "filename": "assets.json",
        "fullPath": true,
        "manifestFirst": true,
        "path": "<PROJECT_ROOT>/test/fake/output/directory",
        "prettyPrint": false,
        "processOutput": [Function],
        "update": false,
      },
      "writer": [Function],
    },
    DefinePlugin {
      "definitions": Object {
        "process.env.DEV_SERVER": false,
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react": "<PROJECT_ROOT>/node_modules/react",
      "react-dom": "<PROJECT_ROOT>/node_modules/react-dom",
      "react-helmet": "<PROJECT_ROOT>/node_modules/react-helmet",
    },
  },
  "target": "web",
  "underwentClientTransform": true,
}
`;
